contracts/RouterStrategy.sol

```diff
@@ -0,0 +1,223 @@
+    }
+
+    function _initializeThis(address _yVault, string memory _strategyName)
+        internal
+    {
+        yVault = IVault(_yVault);
+        strategyName = _strategyName;
+    }
+
+    function name() external view override returns (string memory) {
+        return strategyName;
+    }
+
+    function estimatedTotalAssets() public view override returns (uint256) {
+        return balanceOfWant().add(valueOfInvestment());
+    }
+
+    function prepareReturn(uint256 _debtOutstanding)
+        internal
+        override
+        returns (
+            uint256 _profit,
+            uint256 _loss,
+            uint256 _debtPayment
+        )
+    {
+        uint256 balanceInit = balanceOfWant();
+        _takeVaultProfit();
+        uint256 balanceOfWant = balanceOfWant();
+
+        if (balanceOfWant > balanceInit) {
+            _profit = balanceOfWant.sub(balanceInit);
+        }
+
+        // if the vault is claiming repayment of debt
+        if (_debtOutstanding > 0) {
+            uint256 _amountFreed = 0;
+            (_amountFreed, _loss) = liquidatePosition(_debtOutstanding);
+            _debtPayment = Math.min(_debtOutstanding, _amountFreed);
+            if (_loss > 0) {
+                _profit = 0;
+            }
+        }
+    }
+
+    function adjustPosition(uint256 _debtOutstanding) internal override {
+        if (emergencyExit) {
+            return;
+        }
+
+        uint256 balance = balanceOfWant();
+        if (balance > 0) {
+            _checkAllowance(address(yVault), address(want), balance);
+            yVault.deposit();
+        }
+    }
+
+    function liquidatePosition(uint256 _amountNeeded)
+        internal
+        override
+        returns (uint256 _liquidatedAmount, uint256 _loss)
+    {
+        uint256 balance = balanceOfWant();
+        if (balance >= _amountNeeded) {
+            return (_amountNeeded, 0);
+        }
+
+        uint256 toWithdraw = _amountNeeded.sub(balance);
+        _withdrawFromYVault(toWithdraw);
+
+        uint256 totalAssets = want.balanceOf(address(this));
+        if (_amountNeeded > totalAssets) {
+            _liquidatedAmount = totalAssets;
+            _loss = _amountNeeded.sub(totalAssets);
+        } else {
+            _liquidatedAmount = _amountNeeded;
+        }
+    }
+
+    function _withdrawFromYVault(uint256 _amount) internal {
+        if (_amount == 0) {
+            return;
+        }
+
+        uint256 sharesToWithdraw =
+            Math.min(
+                _investmentTokenToYShares(_amount),
+                yVault.balanceOf(address(this))
+            );
+        if (sharesToWithdraw == 0) {
+            return;
+        }
+
+        yVault.withdraw(sharesToWithdraw, address(this), maxLoss);
+    }
+
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _amountFreed)
+    {
+        (_amountFreed, ) = liquidatePosition(
+            vault.strategies(address(this)).totalDebt
```

Comment:
Could the amount to liquidate be greater than the totalDebt? Maybe ask for `type(uint256).max`?

Reply:
I just checked this.

Vault does the check in https://github.com/yearn/yearn-vaults/blob/master/contracts/Vault.vy#L1081
I also can't do max int because `_investmentTokenToYShares()` does a `mul()`.

Reply:
Could `vault.strategies(address(this)).totalDebt` < `_investmentTokenToYShares(yVault.balanceOf(address(this)))`?

Reply:
Could `vault.strategies(address(this)).totalDebt` < `valueOfInvestment()`?

Reply:
yep, but in that case it will be harvested when `harvest()` is called

Reply:
👍

Reply:
I was thinking the same thing. So there's a chance that all positions would be liquidated and there will still be some profit left in the underlying vault. Wouldn't that make PPS skyrocket on the next harvest?

Reply:
Pretty sure it will not but let me write a test.
I will:
 - harvest
 - send profit to the routed vault
 - set emergency exit
 - harvest
 
 and see what happens 😨

Reply:
isn't it better here to just `yVault.withdraw(yVault.balanceOf(address(this)), address(this), maxLoss)`? Otherwise you're relying on good record keeping by going through the `liquidatePosition` route. Someone could theoretically airdrop some want into this strat and cause some of the yVault shares to remain locked up.

Reply:
found a bug doing this test. well done @plutofeather!

Reply:
Will push new tests soon.